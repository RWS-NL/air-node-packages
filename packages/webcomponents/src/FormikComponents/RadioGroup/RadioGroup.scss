@import "../../styles/vars";

.label {
  color: $textBlack;
  font-size: 0.9rem;
}
