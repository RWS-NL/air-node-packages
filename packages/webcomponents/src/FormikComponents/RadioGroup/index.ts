export {
  default as RadioGroup, RadioGroupProps, Option, OptionValue
} from './RadioGroup';