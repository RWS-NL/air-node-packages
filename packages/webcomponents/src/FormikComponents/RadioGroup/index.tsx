import { FormControlLabel, Radio } from '@material-ui/core';
import { RadioGroup as MUIRadioGroup, RadioGroupProps as MUIRadioGroupProps } from 'formik-material-ui';
import React from 'react';
import css from './RadioGroup.scss';

/** The possible types for the option value */
export type OptionValue = string | number | boolean;

export interface Option<T extends OptionValue> {
  /** The value for the option */
  value: T;
  /** The label shown next to the radio button */
  label: string;
}

export interface RadioGroupProps<T extends OptionValue> {
  /** The options for the radio buttons */
  options: Option<T>[];
}

/**
 * Constructs a radio button group with formik validation
 * @param componentProps Props to pass to the component
 * @example
 * ```jsx
 *  <FastField
 *    name='type'
 *    type='text'
 *    required
 *    data-qa='sample-radio-group'
 * >
 *     {(fieldProps: FieldProps) => (
 *       <RadioGroup<'JOHN'|'CONNOR'>
 *         {...fieldProps}
 *         options={ [{ value: 'JOHN', label: 'John'}, { value: 'CONNOR', label: 'Connor'}] }
 *       />
 *     )}
 * </FastField>
 * ```
 */
export const RadioGroup = <T extends OptionValue>(componentProps: RadioGroupProps<T>) => {
  const { field, form, meta, options, ...props } = componentProps as RadioGroupProps<T> & MUIRadioGroupProps;

  return (
    <MUIRadioGroup field={field} form={form} meta={meta} row {...props}>
      {options.map((option, index) => (
        <FormControlLabel
          key={index}
          value={option.value}
          label={option.label}
          classes={{ label: css.label }}
          labelPlacement='end'
          control={<Radio color='primary' />}
        />
      ))}
    </MUIRadioGroup>
  );
};

export default RadioGroup;
