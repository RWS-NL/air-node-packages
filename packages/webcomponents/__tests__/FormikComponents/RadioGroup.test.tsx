import { FormControlLabel } from '@material-ui/core';
import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import RadioGroup from '../../src/FormikComponents/RadioGroup/RadioGroup';

describe('Component Tests', () => {
  let radioGroup: ShallowWrapper;

  beforeAll(() => {
    radioGroup = shallow(
      <RadioGroup<'JOHN' | 'CONNOR'>
        options={[ { value: 'JOHN', label: 'John' }, { value: 'CONNOR', label: 'Connor' } ]}
      />
    );
  });

  test('GIVEN RadioGroup with 2 options THEN has 2 radio buttons', () => {
    expect(radioGroup.find(FormControlLabel)).toHaveLength(2);
  });
});

describe('Snapshot Testing', () => {
  test('Required Props', () => {
    const radioGroup = shallow(
      <RadioGroup<'JOHN' | 'CONNOR'>
        options={[ { value: 'JOHN', label: 'John' }, { value: 'CONNOR', label: 'Connor' } ]}
      />);
    expect(radioGroup).toMatchSnapshot();
  });
});