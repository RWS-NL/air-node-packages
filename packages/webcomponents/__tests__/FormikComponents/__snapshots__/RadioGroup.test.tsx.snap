// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing Required Props 1`] = `
<FormikMaterialUIRadioGroup
  row={true}
>
  <WithStyles(ForwardRef(FormControlLabel))
    classes={
      Object {
        "label": "label",
      }
    }
    control={
      <ForwardRef(WithStyles)
        color="primary"
      />
    }
    key="0"
    label="John"
    labelPlacement="end"
    value="JOHN"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    classes={
      Object {
        "label": "label",
      }
    }
    control={
      <ForwardRef(WithStyles)
        color="primary"
      />
    }
    key="1"
    label="Connor"
    labelPlacement="end"
    value="CONNOR"
  />
</FormikMaterialUIRadioGroup>
`;
